### PR TITLE
Add support for converters with TypeVars on generic attrs classes

### PR DESCRIPTION
**Pull request by chadrik**
When creating generic classes using `attrs`, converters with type vars are not properly integrated into the generated `__init__`:

```python
from typing import TypeVar, Generic, List, Iterable, Iterator
import attr

T = TypeVar('T')

def int_gen() -> Iterator[int]:
    yield 1

def list_converter(x: Iterable[T]) -> List[T]:
    return list(x)

@attr.s(auto_attribs=True)
class A(Generic[T]):
    x: List[T] = attr.ib(converter=list_converter)
    y: T = attr.ib()

a1 = A([1], 2)   # E: Argument 1 to "A" has incompatible type "Iterator[int]"; expected "Iterable[T]"
```

This MR fixes the bug by copying type vars from the field/attrib into the type extracted from the converter.

### CODE REVIEW

**Comment by chadrik** [comment]
@t4lz @euresti Do either of you have a moment to review this attrs-related change, please?

File: mypy/plugins/attrs.py [hunk 1]
```diff
@@ -115,11 +117,20 @@ def __init__(
...
+                if init_type and self.init_type and self.converter.ret_type:
+                    # The converter return type should be the same type as the attribute type.
+                    # Copy type vars from attr type to converter.
+                    converter_vars = get_type_vars(self.converter.ret_type)
+                    init_vars = get_type_vars(self.init_type)
+                    if converter_vars and len(converter_vars) == len(init_vars):
+                        variables = {
+                            binder.id: arg for binder, arg in zip(converter_vars, init_vars)
+                        }
+                        init_type = expand_type(init_type, variables)
```
Comment (chadrik) [inline mypy/plugins/attrs.py:133]: The idea behind this change is that a converter should ideally return the same type as the attrs field that it is added to.  For example:

```
def int_gen() -> Iterator[int]:
    yield 1

def list_converter(x: Iterable[T]) -> List[T]:
    return list(x)

@attr.s(auto_attribs=True)
class A(Generic[T]):
    x: List[T] = attr.ib(converter=list_converter)
```

`A.x` is type `List[T]` and `list_converter` returns `List[T]`.  In my changes I'm forgoing a complete comparison of the two types, but the assumption tells us that the number and order of TypeVars should be the same.  We use that to generate a replacement map which will replace the converter's broken typevars, with typevars that work within the class's context.

Comment (ikonst) [reply mypy/plugins/attrs.py:136]: (I assume the `def int_gen` slipped in from another example?)

Comment (chadrik) [reply mypy/plugins/attrs.py:136]: yep.  I fixed the example.

**Comment by chadrik** [comment]
Hi @JukkaL, can you or someone else from the mypy team merge this, please?

**Review by hauntsaninja — APPROVED** [review]
Thanks!

**Pull request merged**